www/js/playerlib.js

```diff
@@ -1770,77 +1770,99 @@ function loadLibrary() {
+  // Render
+  renderGenres();
 }
 
 // generate library array
 function filterLib() {
-    allGenres = [];
-    allArtists = [];
-    allAlbums = [];
-    allAlbumsTmp = [];
-    allAlbumCovers = [];
-    allSongs = [];
+  allSongs = fullLib;
 	allSongsDisc = []; // r44g
-    
-    var needReload = false;
-
-    for (var genre in fullLib) {
-	    // AG fix dup artist appearing when Artist assigned to multiple Genres (don't push if already exist)
-        if (allGenres.indexOf(genre) < 0) {
-            allGenres.push(genre);
-        }
-
-        if (LIB.filters.genres.length == 0 || LIB.filters.genres.indexOf(genre) >= 0) {
-            for (var artist in fullLib[genre]) {
-                if (allArtists.indexOf(artist) < 0) {
-                    allArtists.push(artist);
-                }
-
-                if (LIB.filters.artists.length == 0 || LIB.filters.artists.indexOf(artist) >= 0) {
-                    for (var album in fullLib[genre][artist]) {
-						var md5 = $.md5(fullLib[genre][artist][album][0].file.substring(0,fullLib[genre][artist][album][0].file.lastIndexOf('/')));
-						var objAlbum = {'album': album, 'artist': artist, 'compilation': '0', 'imgurl': '/imagesw/thmcache/' + encodeURIComponent(md5) + '.jpg'}; // r44d repl genre with compilation
-                        allAlbumsTmp.push(objAlbum);
-
-                        if (LIB.filters.albums.length == 0 || LIB.filters.albums.indexOf(keyAlbum(objAlbum)) >= 0) {
-                            for (var i in fullLib[genre][artist][album]) {
-                                var song = fullLib[genre][artist][album][i];
-                                song.album = album;
-                                song.artist = artist;
-                                allSongs.push(song);
-                            }
-                        }
-                    }
-                }
-            }
-        }
-    }
-
-    // check filters validity
-    var newFilters = checkFilters(LIB.filters.albums, allAlbumsTmp, function(o) { return keyAlbum(o); });
-
-    if (newFilters.length != LIB.filters.albums.length) {
-        needReload = true;
-        LIB.filters.albums = newFilters;
-    }
-
-    newFilters = checkFilters(LIB.filters.artists, allArtists, function(o) { return o; });
 
-    if (newFilters.length != LIB.filters.artists.length) {
-        needReload = true;
-        LIB.filters.artists = newFilters;
-    }
+  var needReload = false;
+
+  var reduceGenres = function(acc, track) {
+	  (acc[track.genre] = acc[track.genre] || []).push(track);
+	  return acc;
+  };
+
+	var reduceArtists = function(acc, track) {
+  	var artist = track.album_artist || track.artist;
+	  (acc[artist] = acc[artist] || []).push(track);
+	  return acc;
+  };
+
+  var reduceAlbums = function(acc, track) {
+  	(acc[track.album] = acc[track.album] || []).push(track);
+  	return acc;
+  };
+
+  var genreFilter = LIB.filters.genres;
+  if (genreFilter.length) {
+  	allSongs = allSongs.filter(function(track) {
```

Comment:
Doing filtering here on a flat structure will give us more control should we want different sort/filtering views.